### ISSUE TITLE
map_label_to_target should init zero tensor

### ISSUE THREAD

**Issue by ttpro1995** [dasguptar/treelstm.pytorch#7]
Your map_label_to_target for SICK dataset init random tensor. 

```
def map_label_to_target(label,num_classes):
    target = torch.Tensor(1,num_classes) # this is not zero tensor
    ceil = int(math.ceil(label))
    floor = int(math.floor(label))
    if ceil==floor:
        target[0][floor-1] = 1
    else:
        target[0][floor-1] = ceil - label
        target[0][ceil-1] = label - floor
    return target
```

However, in [treelstm](https://github.com/stanfordnlp/treelstm/blob/master/relatedness/TreeLSTMSim.lua) , the author init zero tensor

    local targets = torch.zeros(batch_size, self.num_classes)
    for j = 1, batch_size do
      local sim = dataset.labels[indices[i + j - 1]] * (self.num_classes - 1) + 1
      local ceil, floor = math.ceil(sim), math.floor(sim)
      if ceil == floor then
        targets[{j, floor}] = 1
      else
        targets[{j, floor}] = ceil - sim
        targets[{j, ceil}] = sim - floor
      end

**Comment by dasguptar** [comment]
Hi @ttpro1995 
Good catch! 

I just tried generating some random tensors, and it seems the values are usually quite small, of the order of `1e-37`, although occasionally, things might shoot up to very large values, of the order of `1e+22`. I believe because the random initialisation mostly resulted in small-ish numbers, I was lucky enough to get some reasonable results.

I have made the change to zero initialisation in the `map_label_to_target`function, and am running a job now to see the new results. Will push the new changes as soon as the job finishes. Thanks a ton for spotting the bug!

P.S. Great work porting over the sentiment classification part of the paper as well!

**Comment by Inilien** [comment]
Hi @dasguptar 
sadly there are another few bugs I have found (((
https://github.com/dasguptar/treelstm.pytorch/blob/master/model.py#L57
there is subtraction of 1 from tree.idx. Though I believe you should use just tree.idx instead.

Also, subsequent runs produce different word2index mapping in vocabulary sine you do not sort words and just use plain set with random hash
https://github.com/dasguptar/treelstm.pytorch/blob/master/utils.py#L49
I suggest to use sorted(): 
https://github.com/Inilien/treelstm.pytorch/blob/master/utils.py#L48

I fixed these (and maybe others) in my fork here: https://github.com/Inilien/treelstm.pytorch
But I also switched code to Python 3.
After all those fixes I was able to achieve 0.859269 Pearson coefficient

**Comment by dasguptar** [comment]
Hi, @ttpro1995 @Inilien ,
I fixed the label target tensor initialisation via commit https://github.com/dasguptar/treelstm.pytorch/commit/66a951b545ddae4dd5922e80658982523b03c8b5

Thanks a lot for pointing it out!

**Comment by dasguptar** [comment]
Also, @Inilien , I added commit https://github.com/dasguptar/treelstm.pytorch/commit/ddec16e7968d709ecde4e89cc61b1931193fad4b to fix the indexing bug you found in the model. I was porting from the original Lua code, which is 1 indexed, and probably forgot to account for that!

Regarding the other points:
- I am testing out the fix you suggested to reproduce the same word2index mapping every time. I am not sure whether that affects performance or not, but it would be nice to have reproducibility.
- Also, could you submit a PR with the fixes needed to port the code to Python3? I believe you had submitted PR #8 , but that was too many things in one go. Just a small PR with the Python3 fixes would be highly appreciated! :smile:

**Comment by dasguptar** [comment]
Closing for lack of response, and since most of the issues pointed out here seem to have been fixed!
